### PR TITLE
Add end-to-end tests

### PR DESCRIPTION
**Pull request by ntadimeti**
This PR basically takes the example/run.sh and compares each step's output with an expected golden figure. 

We figured out through this that our models are not exactly reproducible. In another PR, I will work on that and then compare the Step 6 (inference) results. Currently, it's not feasible to test because the bedgraph files are not identical after inference.

### CODE REVIEW

File: tests/end-to-end/run.sh [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+
+echo ""
+echo "Step 7b: Calculate metrics for peak classification after inference..."
+echo ""
+python $root_dir/calculate_baseline_metrics.py \
+    --label_file $out_dir/clean.peaks.bw \
+    --task classification \
+    --test_file $out_dir/inference_latest/no_label_infer_results.h5.peaks.bw \
+    --intervals $out_dir/result.holdout_intervals.bed \
+    --sizes $ref_dir/hg19.auto.sizes \
```
Comment (avantikalal) [inline tests/end-to-end/run.sh]: --sizes does not need to be provided for this command.

File: tests/end-to-end/run.sh [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+echo ""
+echo "Step 7a: Calculate metrics for track coverage after inference..."
+echo ""
+python $root_dir/calculate_baseline_metrics.py \
+    --label_file $data_dir/HSC.80M.chr123.10mb.coverage.bw \
+    --task regression \
+    --test_file $out_dir/inference_latest/no_label_infer_results.h5.track.bw \
+    --intervals $out_dir/result.holdout_intervals.bed \
+    --sizes $ref_dir/hg19.auto.sizes \
```
Comment (avantikalal) [inline tests/end-to-end/run.sh]: --sizes does not need to be provided for this command.

**Comment by ntadimeti** [comment]
rerun tests

File: tests/end-to-end/run.sh [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+env | grep "PATH"
+
+# End-to-end testing script.
+
+verify_diff () {
+    if [ "$3" == "h5" ]
+    then
+        h5diff $1 $2
+    else
+	diff $1 $2
```
Comment (tijyojwad) [inline tests/end-to-end/run.sh]: fix indentation to match g5diff line

File: tests/end-to-end/verify_diff.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        args : parsed argument object.
+
+    """
+    parser = argparse.ArgumentParser(
+        "Utils scripts for verifying identicality of different files.")
+    parser.add_argument('--result_path', type=str,
+                        help='Path to generated result file.')
+    parser.add_argument('--expected_path', type=str,
+                        help='Path to expected result file.')
+    parser.add_argument('--format', type=str,
```
Comment (tijyojwad) [inline tests/end-to-end/verify_diff.py]: I don't see a verify function for `bw`. is that not needed anymore?

File: tests/end-to-end/run.sh [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+# license agreement from NVIDIA CORPORATION is strictly prohibited.
+#
+
+
+set -e
+env | grep "PATH"
+
+# End-to-end testing script.
+
+verify_diff () {
```
Comment (tijyojwad) [inline tests/end-to-end/run.sh]: let's move this to the `verify_diff.py` script as well, so we have all forms of verification encapsulated in a single script.

File: tests/end-to-end/run.sh [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+# Switch to root directory before running script.
+cd $test_root_dir
+
+if [ -d "$out_dir" ]; then
+    rm -rf $out_dir
+fi
+mkdir -p $out_dir
+
+
+echo ""
```
Comment (tijyojwad) [inline tests/end-to-end/run.sh]: let's break each of these steps into separate test.sh scripts with descriptive names. so the individual scripts would look like this - 

```
preamble (a shared setup function to set basic directories)
single test, e.g. Split the given regions of the genome into train, val, and holdout/test intervals...
verify diff and output success or failure
```

File: tests/end-to-end/verify_diff.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        result_path: Path to generated result file.
+        expected_path: Path to expected result file.
+
+    """
+    rank = 0
+    result_model = torch.load(result_path, map_location="cuda:" + str(rank))
+    expected_model = torch.load(expected_path,
+                                map_location="cuda:" + str(rank))
+    for key, _ in result_model['state_dict'].items():
+        val = torch.sum((result_model['state_dict'][key] -
```
Comment (tijyojwad) [inline tests/end-to-end/verify_diff.py]: `val` gets overwritten in each loop of the for. did you mean to use `val +=` here?

File: tests/end-to-end/verify_diff.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+#
+"""Test different output files against expected files."""
+
+import argparse
+
+import sys
+
+import torch
+
+EPS = 0.0001
```
Comment (tijyojwad) [inline tests/end-to-end/verify_diff.py]: since this is global best to add a prefix to it. something like "VERIFY_DIFF_EPS"

File: tests/end-to-end/verify_diff.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+            if diff > EPS:
+                err_msg = "Metrics for " + key + " differ more than " + str(
+                    EPS) + " between result and expected metrics!"
+                raise ValueError(err_msg)
+
+
+args = parse_args()
+if args.format == "torch":
+    verify_torch(args.result_path, args.expected_path)
+if args.format == "text":
```
Comment (tijyojwad) [inline tests/end-to-end/verify_diff.py]: can a user pass both text and torch, or is it one or the other? if it's the latter it would be better to use `if... elif... else` instead of `if... if...`

Comment (ntadimeti) [reply tests/end-to-end/verify_diff.py]: done.

File: tests/end-to-end/verify_diff.py (hunk 8 above)
Comment (ntadimeti) [reply tests/end-to-end/verify_diff.py]: done.

File: tests/end-to-end/run.sh (hunk 3 above)
Comment (ntadimeti) [reply tests/end-to-end/run.sh]: removed fn.

File: tests/end-to-end/verify_diff.py (hunk 4 above)
Comment (ntadimeti) [reply tests/end-to-end/verify_diff.py]: removed, good catch.

File: tests/end-to-end/run.sh (hunk 5 above)
Comment (ntadimeti) [reply tests/end-to-end/run.sh]: done.

File: tests/end-to-end/run.sh (hunk 6 above)
Comment (ntadimeti) [reply tests/end-to-end/run.sh]: done.

File: tests/end-to-end/verify_diff.py (hunk 7 above)
Comment (ntadimeti) [reply tests/end-to-end/verify_diff.py]: done.

File: tests/end-to-end/save_to_h5.sh [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    --batch_size 64 \
+    --prefix $out_dir/no_label \
+    --nolabel
+
+echo ""
+echo "Verifying output against expected results"
+python $test_dir/verify_diff.py --result_path $out_dir/train_data.h5 \
+	              --expected_path $expected_results_dir/train_data.h5 \
+		      --format "h5"
+if [ $? != 0 ]
```
Comment (tijyojwad) [inline tests/end-to-end/save_to_h5.sh]: let's move this block into a function since it's used everywhere. and the function can take in as an argument a custom informative message to print on exit, e.g. "Failed to match train h5" or "Failed to match eval h5" instead of a more generic "Abort test"

File: tests/expected_results/end-to-end/reg_log_7a [hunk 11]
```diff
@@ -0,0 +1,11 @@
+INFO:2020-02-05 17:29:01,552:AtacWorks-calculate-metrics] Loading intervals
```
Comment (tijyojwad) [inline tests/expected_results/end-to-end/reg_log_7a]: can we rename this file to be more descriptive of the content?

File: tests/expected_results/end-to-end/reg_log [hunk 12]
```diff
@@ -0,0 +1,11 @@
+INFO:2020-01-30 14:36:42,958:AtacWorks-calculate-metrics] Loading intervals
```
Comment (tijyojwad) [inline tests/expected_results/end-to-end/reg_log]: same, rename the file to be more descriptive of the content

File: tests/end-to-end/verify_diff.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python
+
```
Comment (tijyojwad) [inline tests/end-to-end/verify_diff.py]: let's move this script to `tests/utils`

File: tests/end-to-end/infer_and_calculate_metrics.sh [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+if [ $? != 0 ]
+then
+    exit
+fi
+echo "Success!"
+
+echo ""
+echo "Calculate metrics for track coverage after inference..."
+echo ""
+python $root_dir/calculate_baseline_metrics.py \
```
Comment (tijyojwad) [inline tests/end-to-end/infer_and_calculate_metrics.sh]: why not split the metrics calculation into a separate end to end test?

File: ci/common/test-pyclaragenomics.sh [hunk 15]
```diff
@@ -19,6 +19,7 @@ cd $PYCLARAGENOMICS_DIR
 
 #Install external dependencies.
 python -m pip install --ignore-installed -r requirements-base.txt && python -m pip install -r requirements-macs2.txt
+apt install hdf5-tools
```
Comment (tijyojwad) [inline ci/common/test-pyclaragenomics.sh]: this works in CI without using `sudo`?

File: tests/end-to-end/save_to_h5.sh (hunk 10 above)
Comment (tijyojwad) [reply tests/end-to-end/save_to_h5.sh]: we can use the abstracted function across all tests

File: tests/end-to-end/train.sh [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+# NVIDIA CORPORATION and its licensors retain all intellectual property
+# and proprietary rights in and to this software, related documentation
+# and any modifications thereto.  Any use, reproduction, disclosure or
+# distribution of this software and related documentation without an express
+# license agreement from NVIDIA CORPORATION is strictly prohibited.
+#
+echo ""
+echo "Train "
+echo ""
+python $root_dir/main.py --train \
```
Comment (tijyojwad) [inline tests/end-to-end/train.sh]: we should still check for non zero exit status

File: tests/expected_results/end-to-end/cla_log [hunk 17]
```diff
@@ -0,0 +1,10 @@
+INFO:2020-01-30 16:54:17,661:AtacWorks-calculate-metrics] Loading intervals
```
Comment (tijyojwad) [inline tests/expected_results/end-to-end/cla_log]: rename file with more descriptive name

File: tests/expected_results/end-to-end/cla_log_7b [hunk 18]
```diff
@@ -0,0 +1,10 @@
+INFO:2020-02-05 17:40:04,642:AtacWorks-calculate-metrics] Loading intervals
```
Comment (tijyojwad) [inline tests/expected_results/end-to-end/cla_log_7b]: same, rename file

File: tests/end-to-end/save_to_h5.sh (hunk 10 above)
Comment (ntadimeti) [reply tests/end-to-end/save_to_h5.sh]: custom messages are printed in python script. So, we don't need specific error messages here.

File: ci/common/test-pyclaragenomics.sh (hunk 15 above)
Comment (ntadimeti) [reply ci/common/test-pyclaragenomics.sh]: this was not supposed to be here, removed it. Added to the dockerfile.

File: tests/end-to-end/infer_and_calculate_metrics.sh (hunk 14 above)
Comment (ntadimeti) [reply tests/end-to-end/infer_and_calculate_metrics.sh]: No reason, just inference and metrics seemed to be 1 step to me.

File: tests/end-to-end/save_to_h5.sh (hunk 10 above)
Comment (ntadimeti) [reply tests/end-to-end/save_to_h5.sh]: Done. verify_diff already prints custom messages. Added a utils.sh script that has a function check_status that can print a custom message too.

File: tests/expected_results/end-to-end/reg_log_7a (hunk 11 above)
Comment (ntadimeti) [reply tests/expected_results/end-to-end/reg_log_7a]: done.

File: tests/expected_results/end-to-end/reg_log (hunk 12 above)
Comment (ntadimeti) [reply tests/expected_results/end-to-end/reg_log]: this isn't needed anymore. removed it.

File: tests/end-to-end/verify_diff.py (hunk 13 above)
Comment (ntadimeti) [reply tests/end-to-end/verify_diff.py]: done.

File: tests/end-to-end/infer_and_calculate_metrics.sh (hunk 14 above)
Comment (ntadimeti) [reply tests/end-to-end/infer_and_calculate_metrics.sh]: done.

File: tests/end-to-end/train.sh (hunk 16 above)
Comment (ntadimeti) [reply tests/end-to-end/train.sh]: agree. removed it by accident. good catch!

File: tests/expected_results/end-to-end/cla_log (hunk 17 above)
Comment (ntadimeti) [reply tests/expected_results/end-to-end/cla_log]: not needed, removed file.

File: tests/expected_results/end-to-end/cla_log_7b (hunk 18 above)
Comment (ntadimeti) [reply tests/expected_results/end-to-end/cla_log_7b]: done.

File: tests/utils/verify_diff.py [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+    parser = argparse.ArgumentParser(
+        "Utils scripts for verifying identicality of different files.")
+    parser.add_argument('--result_path', type=str,
+                        help='Path to generated result file.')
+    parser.add_argument('--expected_path', type=str,
+                        help='Path to expected result file.')
+    parser.add_argument('--format', type=str,
+                        choices=["torch", "text", "h5", "general_diff"],
+                        help='File formats being tested.')
+    parser.add_argument('--metrics', type=str,
```
Comment (tijyojwad) [inline tests/utils/verify_diff.py]: what's this option used for?

File: tests/end-to-end/save_to_h5.sh (hunk 10 above)
Comment (tijyojwad) [reply tests/end-to-end/save_to_h5.sh]: Ah I meant something a little different by the custom messages. verify_diff prints the diff failure between the file names, but sometimes file names get too long and hard to parse. So I was thinking of something like "save to h5 test failed", so in the logs it's easier to follow. we can update it in another PR though. can you file an issue and then resolve this conversation?

Comment (ntadimeti) [reply tests/end-to-end/save_to_h5.sh]: done issue #116

File: tests/utils/verify_diff.py (hunk 19 above)
Comment (ntadimeti) [reply tests/utils/verify_diff.py]: not used anymore, removed.

**Comment by ohadmo** [comment]
rerun tests
